flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannel.java

```diff
@@ -283,10 +283,13 @@ public String toString() {
 	// ------------------------------------------------------------------------
 
 	/**
-	 * Enqueue this input channel in the pipeline for sending unannounced credits to producer.
+	 * Enqueue this input channel in the pipeline for notifying the producer of unannounced credit.
 	 */
 	void notifyCreditAvailable() {
-		//TODO in next PR
+		// We should skip the notification if this channel is already released.
+		if (!isReleased.get() && partitionRequestClient != null) {
```

Comment:
shouldn't we
```
		checkState(partitionRequestClient != null, "Tried to send credit announcement to producer before requesting a queue.");`
```
here as well? At the moment I don't see a valid usecase for `== null` and only a potential problem with the notification not being tried again.

Reply:
In my early implementation, I checked the client state as above. 

Now I change the current way based on two considerations:
1. The credit is always announced after receiving the sender's backlog, so the partition request client should already be initialized before.
2. The check state here only wants to avoid unnecessary announcement after the client is closed because of task failure or network exception. In this case, the closed client is in expectation and we do not want to throw disturbed messages by `checkState`.

Reply:
But neither `CreditBasedClientHandler#ctx`, nor `RemoteInputChannel#partitionRequestClient` are set to `null` when the channel is closed. They are only set once in `CreditBasedClientHandler#channelActive` and `RemoteInputChannel#requestSubpartition` and then the channel is in a valid state to interact with. Or am I not seeing something else here?

Reply:
Yes, the partition client is not set null explicitly when the channel is closed. But I am wondering it should be set null to avoid the released scenario.

I will restore the `checkState` for the client as you suggested and it seems more reasonable currently.

Reply:
If we set it to `null` concurrently, we may need some more synchronisation here, so I'd leave that logic as it is for now.

Reply:
agree